Upgrade to Spring Boot 2.5.3 and Spring Cloud 2020.0.3

gicappa opened the pull request:
Bumping a new **spring-boot/cloud** version

Found some _minor incompatibilities_ due to the evolution of the spring-boot/data platform from the **2.4.x** to the **2.5.x**

- The `Pageable` interface has a new method `withPage()` that has been implemented by delegating to the inner peageable delegate object implementation.
- `QueryRepositoryConfig` and `RepositoryRestConfig` are inheriting from `RepositoryRestConfigurer` that changed the signature of the method `configureRepositoryRestConfiguration()` adding the `CorsRegistry` parameter. The signature has been updated but the parameter has not been taken into consideration.
- In `ModelingRestExceptionHandler` the `DefaultErrorAttributes` class changed the signature of `getErrorAttributes()` switching from a boolean param to include the stack trace to a class `ErrorAttributeOptions` to manage several kind of error attributes to be included.

gicappa committed:
GH-3688 Bumbed a new version of spring-cloud and spring-boot

To test the activiti-cloud build in isolation I used the preview
mechanism specifying to use the Activiti preview version:

0.0.1-dev-gicappa-GH-3688-preview-211-SNAPSHOT

gicappa committed:
GH-3688 Bumping a new spring-boot/cloud version

Found some minor incompatibilities due to the evolution of the spring-boot/data platform from the 2.4.x to the 2.5.x

- The Pageable interface has a new method withPage() that has been implemented by delegating to the inner peageable implementation.

- QueryRepositoryConfig and RepositoryRestConfig are inheriting from RepositoryRestConfigurer that changed the signature of the method configureRepositoryRestConfiguration() adding the CorsRegistry parameter. The signature has been updated but the parameter has not been taken into consideration.

- In ModelingRestExceptionHandler the DefaultErrorAttributes class changed the signature of getErrorAttributes() switching from a boolean param to include the stack trace to a class ErrorAttributeOptions to manage several kind of error attributes to be included.

gicappa committed:
GH-3688 Bumping a new spring-boot/cloud version

FIX: with the new spring-boot version the test class QueryApplicationEntityIT was failing reporting a missing Id in the list of ApplicationEntity returned by the test.
The QueryRepositoryConfig class, the the ApplicationEntity was not listed with the classes allowed to expose their ids.

gicappa committed:
GH-3688 Bumping a new spring-boot/cloud version

FIX: with the new spring-boot version the test class QueryAdminProcessServiceTasksIT was failing on the shouldGetServiceTaskById.
In the QueryRepositoryConfig class, the ServiceTaskEntity was not listed with the classes allowed to expose their ids.

gicappa committed:
FIX: database initialization in sb 2.5.2 has changed causing a bug in the tests.

It has been added the properties spring.jpa.defer-datasource-initialization=true to the application-test.properties file to be retro-compatible with the previous version.

reference https://github.com/spring-projects/spring-boot/wiki/Spring-Boot-2.5-Release-Notes#hibernate-and-datasql

gicappa committed:
Factorized the fix introduced in the spring-boot version 2.5.3

gicappa committed:
GH-3688 FIX wrong position of a field / static code analysis

erdemedeiros commented on activiti-cloud-runtime-bundle-service/pom.xml:
```diff
@@ -8,23 +8,42 @@
   </properties>
   <dependencyManagement>
     <dependencies>
+      <!-- 26/07/2021
+After upgrading spring-boot from 2.5.2 to 2.5.3, the
+build started failing because it was not anymore able
+to create the ApplicationContext.
+The error reported was:
+
+Caused by: org.springframework.context.ApplicationContextException: Failed to start bean 'inputBindingLifecycle'; nested exception is java.lang.NoSuchMethodError: org.springframework.integration.amqp.inbound.AmqpInboundChannelAdapter.<init>(Lorg/springframework/amqp/rabbit/listener/AbstractMessageListenerContainer;)V
+at org.springframework.context.support.DefaultLifecycleProcessor.doStart(DefaultLifecycleProcessor.java:181)
+
+The issue seems to be caused by this change in spring-integration-amqp:
+
+https://github.com/spring-projects/spring-integration/commit/76f77ccd2c5f94e000495357e66c4f99fffc6010
+
+Since the issue happens only due to the dependencies imported for testing it has been fixed rolling
+back the library spring-integration-amqp the the previous version (5.5.1)
+-->
+      <dependency>
+        <groupId>org.springframework.integration</groupId>
+        <artifactId>spring-integration-amqp</artifactId>
+        <version>5.5.1</version>
+      </dependency>
```
You should be able to remove this, as it's already declared in `activiti-cloud-build-dependencies-parent`.

gicappa replied:
My first implementation was relying on the definition declared in `activiti-cloud-build-dependencies-parent` but `activiti-cloud-runtime-bundle-service` was failing again. I tried to inspect on the possible dependency issue but the complexity of the dependencies where preventing me to do it in a proper time and was diverting me from the primary goal of the pull request.

gicappa replied:
Fixed by the commit 559fbc1

erdemedeiros commented on activiti-cloud-service-common/activiti-cloud-services-dbp-rest/src/main/java/org/activiti/cloud/alfresco/argument/resolver/AlfrescoPageRequest.java:
```diff
@@ -82,6 +82,11 @@ public AlfrescoPageRequest first() {
 
+    @Override
+    public Pageable withPage(int pageNumber) {
+        return pageable.withPage(pageNumber);
```
Actually, after double-checking, what you need to set here is the `skipCount` with `pageNumber * getPageSize()` because `getPageNumber` is overridden [here](https://github.com/Activiti/activiti-cloud/pull/455/files#diff-11ec11e53366eb734a751267e5ae0caea469ef247150dcab759642f9b2c891a1R51). Please, add a new test to  `AlfrescoPageRequestTest` to cover this.

gicappa replied:
Fixed as you suggested, thanks.

erdemedeiros reviewed (changes requested):
Two small details. LGTM otherwise.

gicappa committed:
GH-3688 Aligning the test class to JUnit5

Minor refactorings to align to JUnit5 standards and to improve consistency through the code.

gicappa committed:
GH-3688 Implemented AlfrescoPageRequest.withPage behavior

The method withPage has been introduced by the new spring boot version. The previous implementation was not taking into account the definition of skipCount and maxItem of the AlfrescoPageRequest class itself and was erroneously relying on the delegated Pageable class.

The method has been implemented calculating the page offset by multiplying the page size times the desired page number.

gicappa committed:
GH-3688 Factorised the rollback of the spring-integration version to 5.5.1

The dependency to version 5.5.1 was introduced twice since the
activiti-cloud-runtime-bundle-service/pom.xml was receiving the rollback
of the spring-integration-amqp to the 5.5.1 version from the
activiti-cloud-build-dependencies-parent/pom.xml.
Overriding the whole spring-integration version using the bom in the
activiti-cloud-build-dependencies-parent/pom.xml solved the issue
(thanks to Elias for the hint).

gicappa committed:
Bump of the new released of spring-boot 2.5.3

gicappa committed:
FIX: spring-integration-amqp raises an exception creating appcontext

After upgrading spring-boot from 2.5.2 to 2.5.3, the
build started failing because it was not anymore able
to create the ApplicationContext.

The error reported was:

Caused by: org.springframework.context.ApplicationContextException: Failed to start bean 'inputBindingLifecycle'; nested exception is java.lang.NoSuchMethodError: org.springframework.integration.amqp.inbound.AmqpInboundChannelAdapter.<init>(Lorg/springframework/amqp/rabbit/listener/AbstractMessageListenerContainer;)V
	at org.springframework.context.support.DefaultLifecycleProcessor.doStart(DefaultLifecycleProcessor.java:181)

The issue seems to be caused by this change in spring-integration-amqp:
https://github.com/spring-projects/spring-integration/commit/76f77ccd2c5f94e000495357e66c4f99fffc6010

Since the issue happens only due to the dependencies imported for testing it has been fixed rolling
back the library spring-integration-amqp the the previous version (5.5.1)

erdemedeiros reviewed (approved):
LGTM!

erdemedeiros commented:
Ref https://github.com/Activiti/Activiti/issues/3688
